pkg/operators/klusterlet/controllers/bootstrapcontroller/bootstrapcontroller.go

```diff
@@ -0,0 +1,158 @@
+	corelister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+)
+
+// bootstrapController watches bootstrap hub kubeconfig secret, if the secret is changed with hub kube-apiserver ca or apiserver
+// endpoints, this controller will make the klusterlet re-bootstrap to get the new hub kubeconfig from hub cluster by deleting
+// the current hub kubeconfig secret and restart the klusterlet agents
+type bootstrapController struct {
+	kubeClient       kubernetes.Interface
+	klusterletLister operatorlister.KlusterletLister
+	secretLister     corelister.SecretLister
+}
+
+// NewBootstrapController returns a bootstrapController
+func NewBootstrapController(
+	kubeClient kubernetes.Interface,
+	klusterletInformer operatorinformer.KlusterletInformer,
+	secretInformer coreinformer.SecretInformer,
+	recorder events.Recorder) factory.Controller {
+	controller := &bootstrapController{
+		kubeClient:       kubeClient,
+		klusterletLister: klusterletInformer.Lister(),
+		secretLister:     secretInformer.Lister(),
+	}
+	return factory.New().WithSync(controller.sync).
+		WithInformersQueueKeyFunc(helpers.BootstrapSecretQueueKeyFunc(controller.klusterletLister), secretInformer.Informer()).
+		ToController("BootstrapController", recorder)
+}
+
+func (k *bootstrapController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	queueKey := controllerContext.QueueKey()
+	if queueKey == "" {
+		return nil
+	}
+
+	keys := strings.Split(queueKey, "/")
+	if len(keys) != 2 {
+		// this should not happen, do nothing
+		return nil
+	}
+	klusterletNamespace := keys[0]
+	klusterletName := keys[1]
+
+	klog.V(4).Infof("Reconciling bootstrap hub kubeconfig secret %q", klusterletNamespace+"/"+helpers.BootstrapHubKubeConfig)
+
+	bootstrapHubKubeconfigSecret, err := k.secretLister.Secrets(klusterletNamespace).Get(helpers.BootstrapHubKubeConfig)
+	switch {
+	case errors.IsNotFound(err):
+		// the bootstrap hub kubeconfig secret not found, do nothing
+		return nil
+	case err != nil:
+		return err
+	}
+
+	bootstrapKubeconfig, err := k.loadKubeConfig(bootstrapHubKubeconfigSecret)
+	// a bad bootstrap secret, ignore it
+	if err != nil {
+		controllerContext.Recorder().Warningf("BadBootstrapSecret",
+			fmt.Sprintf("unable to load hub kubeconfig from secret %q: %v", klusterletNamespace+"/"+helpers.BootstrapHubKubeConfig, err))
+		return nil
+	}
+
+	hubKubeconfigSecret, err := k.secretLister.Secrets(klusterletNamespace).Get(helpers.HubKubeConfig)
+	switch {
+	case errors.IsNotFound(err):
+		// the hub kubeconfig secret not found, could not have bootstrap yet, do nothing
+		return nil
+	case err != nil:
+		return err
+	}
+
+	hubKubeconfig, err := k.loadKubeConfig(hubKubeconfigSecret)
+	// the hub kubeconfig secret has errors, reload the klusterlet agents
+	if err != nil {
+		return k.reloadAgents(ctx, controllerContext, klusterletNamespace, klusterletName)
```

Comment:
I‘d rather we do not reload here. An  empty hubkubeconfig secret can also trigger the reload, is it?

Reply:
do we have an empty hubkubecofnig? I want to add a judgement to judge the placeholder secret before here

Reply:
return here directly